src/ocean/meta/types/Function.d

```diff
@@ -0,0 +1,143 @@
+
+module ocean.meta.types.Function;
+
+import ocean.meta.AliasSeq;
+import ocean.meta.traits.Basic;
+import ocean.meta.types.Typedef;
+
+/*******************************************************************************
+
+    Evaluates to a compile-time sequence representing the parameters of Fn
+
+    Params:
+        Callable = callable type (function/delegate/function pointer) or
+            function symbol
+
+*******************************************************************************/
+
+public template ParametersOf ( Callable... )
+{
+    static assert (Callable.length == 1);
+
+    static if (is(Callable[0] Params == function))
+        alias AliasSeq!(Params) ParametersOf;
+    else static if (is(typeof(Callable[0]) Params == function))
+        alias AliasSeq!(Params) ParametersOf;
+    else static if (is(Callable[0] TDgPtr == delegate))
+        alias AliasSeq!(ParametersOf!(TDgPtr)) ParametersOf;
+    else static if (is(Callable[0] TFunc == TFunc*))
+        alias AliasSeq!(ParametersOf!(TFunc)) ParametersOf;
+    else static if (isTypedef!(Callable[0]))
+        alias ParametersOf!(TypedefBaseType!(Callable[0])) ParametersOf;
+    else
+        static assert (false, "Argument is not a function");
+}
+
+///
+unittest
+{
+    void foo(int x, ref double y, char[] z) {}
+    alias ParametersOf!(typeof(foo)) Params;
+
+    static assert (Params.length == 3);
+    static assert (is(Params[0] == int));
+    static assert (is(Params[1] == double));
+    static assert (is(Params[2] == char[]));
+}
+
+unittest
+{
+    void foo(int x, ref double y, char[] z) {}
+    alias ParametersOf!(typeof(&foo)) Params;
+
+    static assert (Params.length == 3);
+    static assert (is(Params[0] == int));
+    static assert (is(Params[1] == double));
+    static assert (is(Params[2] == char[]));
+}
+
+unittest
+{
+    void delegate(int x, ref double y, char[] z) dg;
+    alias ParametersOf!(typeof(dg)) Params;
+
+    static assert (Params.length == 3);
+    static assert (is(Params[0] == int));
+    static assert (is(Params[1] == double));
+    static assert (is(Params[2] == char[]));
+}
+
+/*******************************************************************************
+
+    Params:
+        Callable = callable type (function/delegate/function pointer) or
+            function symbol
+
+    Returns:
+        evaluates to return type of the callable
+
+*******************************************************************************/
+
+public template ReturnTypeOf ( Callable... )
+{
+    static assert (Callable.length == 1);
+
+    static if (is(typeof(Callable[0]) == function))
+    {
+        alias typeof({
+                ParametersOf!(Callable[0]) args;
+                return Callable[0](args);
+            } ()) ReturnTypeOf;
+    }
+    else
+    {
+        alias typeof({
+                ParametersOf!(Callable[0]) args;
+                Callable[0] call;
+                return call(args);
+            } ()) ReturnTypeOf;
+    }
```

Comment:
Couldn't this be done using `static if (is(Callable[0] R == return))`?

Reply:
Only if it is a function AFAIR (so, for example, won't work for `opCall` object)